### ISSUE TITLE
Reviving Ratchet: Let's give Ratchet the love it deserves!

### ISSUE THREAD

**Issue by clue** [ratchetphp/Ratchet#1054]
Ratchet is a great piece of software and it is successfully used in countless projects! But let's be realistic: It could use some love!

At the moment, we're making very good progress with ReactPHP v3 (see e.g. [reactphp/event-loop#271](https://github.com/reactphp/event-loop/issues/271)) and Framework X. As a next step, we want to achieve the same level of velocity for Ratchet. Supporting the latest PHP and PHPUnit versions feels like a good starting point to bring Ratchet back to speed again, followed by tackling the huge pile of open support requests. So in a gist, we know what to do with Ratchet, the problem is that we have to find the time to actually do it. I've also reached out to the other maintainers, but before spending too much time on the details, I would like to see how much demand there is to work out a sustainable plan for the longer-term development.

As a first step, help us figure out demand by leaving a comment or an emoji reaction:

* 👍 Like the idea, generally interested
* ❤️ Love the idea and would consider sponsoring further development of Ratchet
* 🚀 Ratchet is used in our company (let's have a chat!)
* Anything else, please join the discussion!

**Comment by dvdknaap** [comment]
So whats the plan because this package is now blocking to upgrade to symfony 7 ?

Can we upgrade to the latest version and then do a release and then work on other features ?

**Comment by SimonFrings** [comment]
@dvdknaap That is ultimately the plan. Like @clue stated above, we're currently looking into project funding and once we know how to proceed with Ratchet, we'll probably start by adding support for the latest PHP and PHPUnit versions. We also have a separate discussion for Symfony 7 in #1045, so I think it makes sense to discuss further details in there.

@clue and the rest of our team are currently busy working on the release for a first version of [Framework X](https://github.com/clue/framework-x), so we'll come back here once we freed some time :+1:

**Comment by userofjack** [comment]
I have enough time to participate in the development of Ratchet, and we are currently using it. How should I participate in development?

**Comment by neoteknic** [comment]
@clue maybe add write access to people who want to help ? like @userofjack

**Comment by igorbelikov** [comment]
I think that many people would support this post, but no one knows about it, because everyone is used to seeing the message “last modified 3 years ago” and simply do not know that you are going to continue supporting

overall this is great news and we are still using ratchet

just start making changes or apply pull requests :)

**Comment by rperdomoalhc** [comment]
> I think that a lot of people would support this post, but no one knows about it, because everyone is used to seeing the message “last modified 3 years ago” and simply do not know that you are going to continue supporting
> 
> overall this is great news and we are still using ratchet
> 
> just start making changes or apply pull requests :)

Since I saw the last commit, I thought it was abandoned.

I'm deciding to integrate it into my company. They did a very good job!

**Comment by johnrix** [comment]
> Since I saw the last commit, I thought it was abandoned.

Ditto. Perhaps update the README with a reference to this thread. That will then also reset the last commit time.

**Comment by clue** [comment]
Thank you so much for all the amazing feedback we've received on Ratchet! We are truly grateful for the excitement and the offers to help with the development. Excited to use this momentum to move forward! 🚀

1. First we need to get the project up to date with the latest versions, so this is something I personally would like to work on first. Once this is completed, I would love to get more people involved, so we greatly appreciate any help and contributions.

2. Right now, we are finalizing the project handover with @cboden and hope to get the project up to speed again as soon as possible. In the meantime, we also work on a number of other bigger open-source projects like @ReactPHP (which Ratchet builds on top of) and need to balance our priorities and find adequate project funding to allow us to keep shipping.

3. You can help us! For now, I would like to set up a simple sponsorship goal of reaching 10 new sponsors (any amount) to see how much the community is willing to support this project. Once reached, we will start pushing the next PRs to get Ratchet updated to support the latest PHP and PHPUnit versions and use this as a starting point for bigger updates to come. Even just 10 x $1 would help achieve our initial goal, but the more sponsors the better obviously, so especially curious about any company sponsors reaching out!

Please help us getting Ratchet up to speed by [becoming a sponsor through ReactPHP](https://github.com/sponsors/reactphp), we're thankful for every contribution! ❤️

**Comment by clue** [comment]
Also agree it's a good idea to link more prominently to this ticket, I've updated the readme for the next couple of weeks and will continue posting more updates as soon as possible!

**Comment by newmediacrew** [comment]
I just cancelled an old GitHub pro account that was never used. Using that money to sponsor this one.

**Comment by whoisninjaturtle** [comment]
Sorry to cross over but https://github.com/ratchetphp/Pawl is in exactly the same boat. My company uses Pawl specifically and it's to the point where we're ready to maintain our own fork instead if that project is truly unmaintained. To that effect I can participate in development on either repo but would like to get an idea of what's going on with Pawl.

**Comment by mkrasselt1** [comment]
> Sorry to cross over but https://github.com/ratchetphp/Pawl is in exactly the same boat.
May be it's a good idea to join source an make a common version that's client and server in the same project?

**Comment by whoisninjaturtle** [comment]
The existing repository structure makes sense; RFC6455 defines agnostic components that anyone could use in their Server/Client implementations. These components are separate from the implementations themselves. Pawl is purely an issue of not having any maintainers at the moment.

**Comment by mkrasselt1** [comment]
> Please help us getting Ratchet up to speed by [becoming a sponsor through ReactPHP](https://github.com/sponsors/reactphp), we're thankful for every contribution! ❤️
@clue whats the progress on that :) ?

**Comment by wpjscc** [comment]
I look forward to it becoming part of the reactphp middleware, so that it can work together with reactphp/http or framework-x. It is very useful to provide both http service and websocket service on the same port, especially when http and websocket are related, as they are in the same process.

**Comment by dvdknaap** [comment]
looks like this project is still dead. does anyone has a replacement ?

**Comment by dvdknaap** [comment]
still no update after all those months. they are only hidding replies and collecting money.

**Comment by danielgoncalves1976** [comment]
Hello, is there any progress on the revivification of Ratchet? I would like to participate and contribute.

**Comment by treck** [comment]
I'm very much appreciating the decission to bring ratchet up to new life, as i'm planning to use it in the future and was a bit worried about the staleness.
Would also like to support whereever possible. How can i?

**Comment by mkrasselt1** [comment]
I needed to upgrade my project using ratchet to php8.3 and ran into issues again - can we prioritize requirements upgrades/deprecation first? would work on/provide a pull request (without errors) if that helps... yet i am not that good at testing....i would love to help on this!

**Comment by elerocks** [comment]
Since one year has almost passed, I think it is appropriate to ask again whether the project is abandoned and we should move to another WS library. Currently I've postponed the update of my Laravel project just because of the Ratchet, but I cannot do that indefinitely. 

Either answer is completely OK, but no answer keeps us in the dark with the planning. Thank you!

**Comment by WyriHaximus** [comment]
> Either answer is completely OK, but no answer keeps us in the dark with the planning. Thank you!

Without going into details as it isn't my place to, @clue has been having a lot on his place the past couple of years. This is also why ReactPHP v3 didn't get as much movement as we both wanted. That said I do have a call with him in a month now that things are falling into place on his end, also have a call next week with someone else working for him to catch up. This will be one of the topics on the agendas.

**Comment by dvdknaap** [comment]
> > Either answer is completely OK, but no answer keeps us in the dark with the planning. Thank you!
> 
> Without going into details as it isn't my place to, [@clue](https://github.com/clue) has been having a lot on his place the past couple of years. This is also why ReactPHP v3 didn't get as much movement as we both wanted. That said I do have a call with him in a month now that things are falling into place on his end, also have a call next week with someone else working for him to catch up. This will be one of the topics on the agendas.

If you want to speed up the development process. Please take a look at https://github.com/ratchetphp/Ratchet/pull/1014 
His pull request is working with php 8 and symfony 7.

I've been testing his update for months now in production without any issues.

**Comment by zhiyong-ft** [comment]
@dvdknaap thanks for the pointer. I will also use this fork to upgrade for the time being.

**Comment by WyriHaximus** [comment]
> also have a call next week with someone else working for him to catch up. This will be one of the topics on the agendas.

Small update on this, just got out of that call, and something is coming soon.

**Comment by whoisninjaturtle** [comment]
This has to be incredibly frustrating for all the people with pull requests languishing for years. Can we at least confirm at this point that this is not a financial issue, given that support subscriptions were described as prerequisite to this reboot?

**Comment by arcanisgk** [comment]
Roadmap???

**Comment by Tylemagne** [comment]
Trying to get TLS to work this past week has made me pretty confident that Ratchet's documentation and example code could really REALLY use some work. It's "switch to Python or C# for WebSockets" bad.

**Comment by mkrasselt1** [comment]
> Trying to get TLS to work this past week has made me pretty confident that Ratchet's documentation and example code could really REALLY use some work. It's "switch to Python or C# for WebSockets" bad.

i would recommend putting ratchet behind a reverse proxy of nginx or similar software

**Comment by arcanisgk** [comment]
> i would recommend putting ratchet behind a reverse proxy of nginx or similar software

i use nginx reverse proxy in docker

**Comment by mkrasselt1** [comment]
So, whats the project status now?
My project just failed in production due to the distro setting its php version above 8.2 
and i had to switch to one of the projects pull requests. I guess this project has lots of developers backing it up an fixing it, but i wonder if an when there will be a new release (even just some housekeeping) at all!

**Comment by dvdknaap** [comment]
> So, whats the project status now? My project just failed in production due to the distro setting its php version above 8.2 and i had to switch to one of the projects pull requests. I guess this project has lots of developers backing it up an fixing it, but i wonder if an when there will be a new release (even just some housekeeping) at all!

They only want sponsors they still don't know anything.

Check my message earlier if you want php 8 support

**Comment by clue** [comment]
> **Next Steps**
With the sponsorship goal reached, we'll finally be able to start reviving Ratchet in April. As mentioned previously, our first priority is to get the project up to date with the latest versions of PHP and PHPUnit, as well as addressing the backlog of support requests.

**Progress update:** Over the past week, I’ve dedicated significant time to updating Ratchet as promised. The project is now compatible with the latest PHP and PHPUnit versions locally, and I’ve started addressing various incompatibilities with upstream components. The first PR is now open (#1088), along with a rough roadmap of upcoming PRs. If time allows, I plan to post a more detailed update and roadmap next week.

>  […] I understand everybody's excited about moving forward with Ratchet […]

We're equally excited to move forward with Ratchet and will do so as fast as humanly possible. I've shared some very personal reasons above for why things may not always be able to move as fast as we'd all hope, but I will continue to do my best to help move forward with this project as much as I can.

To ensure Ratchet’s continued progress (and our own wellbeing), I'm sure you understand we reserve the right to block users who create distractions or disrupt the project — whether intentionally or not. Please keep discussions constructive and on-topic.

Currently, the project receives about $120/month in sponsorships, which I think is a great first step! As you can imagine, this only covers a small fraction of the time and resources needed for professional support and development for a project of this size. If you’d like to help support Ratchet’s future, please consider sponsoring the project or reaching out — every bit of support truly helps. ❤️

Thank you all for your understanding, encouragement, and patience as we work to make Ratchet better for everyone.

**Comment by clue** [comment]
**Major update**: I’m excited to share some great progress on Ratchet! Over the past few weeks, we’ve put in a ton of work merging PRs, modernizing the codebase, improving PHP compatibility, supporting latest PHP and Symfony versions, and getting our CI pipeline running smoothly again. It’s been a huge effort, and I think we all agree Ratchet is in a much better state now.

**Next** on my list is cleaning up the backlog of support requests (thanks for hanging in there if you’re waiting on something). I’ve got a busy week coming up next week, but I’ll do my best to continue working on Ratchet as soon as time allows again.

**On that note**: Sponsorships really do make a difference. More sponsors directly translate to faster releases and more features. The more support we get, the quicker I can shift some other commitments around and focus on getting these releases out. It's as simple as that. If you're using Ratchet in production or getting value from it, please consider [sponsoring ReactPHP](https://github.com/sponsors/reactphp). ❤️

**Upcoming**: More updates coming when I get a chance in two weeks. Hoping to have more good news for you all soon! 🎉

**Comment by clue** [comment]
Hey everyone,

Thanks for your patience and support as we've been working through Ratchet's backlog. Since our last update, we've put tons of time and effort into this, and honestly, we're pretty proud of where we've gotten Ratchet to. We've fixed many bugs and compatibility issues that Ratchet is now at a really solid point and *close* to a release!

**What we've accomplished so far**

- Closed around 170 out of 180 open issues and closed all open PR's like announced in [#1100](https://github.com/ratchetphp/Ratchet/issues/1100)
- Update test environment to fix running tests on GitHub Actions ([#1088](https://github.com/ratchetphp/Ratchet/pull/1088))
- Rename master branch to 0.4.x and add installation instructions ([#1091](https://github.com/ratchetphp/Ratchet/pull/1091))
- Update tests to support PHPUnit ^7.5 and psr/http-message ^2.0 on PHP 7.2+ ([#1092](https://github.com/ratchetphp/Ratchet/pull/1092))
- Update tests to support PHPUnit ^9.6 ([#1093](https://github.com/ratchetphp/Ratchet/pull/1093))
- Full support for PHP 8.1 and PHP 8.0 ([#1094](https://github.com/ratchetphp/Ratchet/pull/1094))
- Update to properly support Symfony 6 (PHP 8+) ([#1095](https://github.com/ratchetphp/Ratchet/pull/1095))
- Support PHP 8.2 and work around reporting dynamic properties ([#1096](https://github.com/ratchetphp/Ratchet/pull/1096))
- Support PHP 8.3 and work around unserialize warnings ([#1097](https://github.com/ratchetphp/Ratchet/pull/1097))
- Update to support Symfony 7 (PHP 8.2+) ([#1098](https://github.com/ratchetphp/Ratchet/pull/1098))
- Support PHP 8.4 and work around reporting implicitly nullable types ([#1099](https://github.com/ratchetphp/Ratchet/pull/1099))
- Merged important cleanup work like the funding docs ([#1101](https://github.com/ratchetphp/Ratchet/pull/1101)) and issue templates ([#1102](https://github.com/ratchetphp/Ratchet/pull/1102))
- Got discussions running ([#1103](https://github.com/ratchetphp/Ratchet/pull/1103)) to make it easier for everyone to contribute
- Update to use global default loop with react/event-loop v1.2+ ([#1105](https://github.com/ratchetphp/Ratchet/pull/1105))
- Update to use new SocketServer API with react/socket v1.9+ ([#1106](https://github.com/ratchetphp/Ratchet/pull/1106) & [#1107](https://github.com/ratchetphp/Ratchet/pull/1107))
- Assign empty remote address if socket address is unknown (Unix sockets) on PHP 8.1+ ([#1108](https://github.com/ratchetphp/Ratchet/pull/1108))
- Fix warning when handling HTTP request with missing Origin header ([#1109](https://github.com/ratchetphp/Ratchet/pull/1109))
- Fix rejecting invalid HTTP request starting with newlines ([#1110](https://github.com/ratchetphp/Ratchet/pull/1110))

**What's next**

While working on reviving Ratchet, we've also begun exploring a broader vision for modern WebSocket development in PHP. This has led to early work on Sprocket, a next-generation WebSocket toolkit that builds on the lessons learned from Ratchet's development and success. Though still in early stages, Sprocket represents what I believe could be a significant step forward for the ecosystem, with detailed plans and vision outlined in the community discussion in #1111 for those interested in following the project's longer-term direction.

**Growing the Ecosystem**

Thanks to community support, we've been able to maintain and significantly improve Ratchet. For ambitious projects like Sprocket that could benefit the entire PHP WebSocket ecosystem, we're looking for partners who share our vision. If your company relies on real-time web technologies and wants to help shape the future of PHP WebSockets, we'd love to discuss how we can work together. If you'd like to support us, please reach out via email or consider [sponsoring the project](https://github.com/sponsors/reactphp) – your support really makes a difference. ❤️

We'll keep you updated on both Ratchet's ongoing development and Sprocket's progress as we move forward. Really appreciate everyone who's been part of this community and helped get us this far.

**Comment by elerocks** [comment]
Hello, forgive me for asking, but lastly you've mentioned that you're close to release. It's been few months, can you give us an estimate please?

**Comment by frenci8** [comment]
Hi,
Could you please provide an update on the status of the new release? Thank you.

**Comment by TheDigitalOrchard** [comment]
Is it correct that $111,173.75 USD has now been raised in total? That's a lot money to see no new releases on this project in more than a year. Some of us just need dependency-compatibility updates, which are relatively minor and not related to the v3 release. Thanks for your attention to this.

**Comment by clue** [comment]
Hey everyone! Back with an update after a rough few months, and some really exciting news to share.

I've recently taken on a new CTO role where ReactPHP and Framework X power a substantial part of our production infrastructure, which means continued investment in the ecosystem is now part of my day job. We're actively hiring senior software engineers who are excited about shipping cool stuff with this stack. If that sounds interesting, check out my profile (@clue) and reach out!

On the ecosystem funding side: Those following ReactPHP's OpenCollective may have noticed a *significant* new contribution. I'll share full details in a proper announcement soon, but this is genuinely game-changing and will help move things forward substantially. All numbers are public: Ratchet-specific sponsorship remains modest at ~$150-200/month, but improvements to the foundation benefit everything built on top of it, including Ratchet. 🎉

On the personal side: December was rough. Hospital stay for a life-threatening health issue, plus a move. I'm recovering and getting back on track. Life sometimes has other plans I guess.

Current focus is ReactPHP v3 and Framework X, which directly benefit Ratchet since it builds on top of them. I remain committed to Ratchet and definitely plan to continue when time allows. But I'm not going to burn myself out or apologize for having a life outside GitHub. The progress earlier in 2025 was substantial: we closed ~170 issues, added PHP 8.4 and Symfony 7 support, and got the project to a solid state. That work doesn't disappear because I've been quiet for a few months.

More updates soon. Thanks to everyone who's been supportive, it really matters! Constructive contributions and patience are always welcome. Comments that aren't productive will continue to be marked off-topic.

**Comment by sargearmstrong** [comment]
Glad you are alive friend, life is nuts. Thank you for the update, and for taking care not to burn out!
